### PR TITLE
feat(jetbrains): show profile balance details

### PR DESCRIPTION
**Pull request by johnnyeric**
## Context

Bring the JetBrains profile balance UI in line with the VS Code balance UI now that `main` exposes Kilo Pass data from the core API.

## Implementation

Maps `kiloPass` from the backend profile response into the shared JetBrains RPC DTO, normalizes the nullable OpenAPI field, and renders personal-account Kilo Pass usage, bonus credits, renewal date, Dashboard, and Top up actions in the JetBrains profile settings UI. Org accounts continue to hide Kilo Pass details.

## Screenshots / Video

jetbrains

### Manual/local verification

- Agent: `env TMPDIR=/private/tmp JAVA_HOME=/opt/homebrew/opt/openjdk@21 GRADLE_USER_HOME=/private/tmp/kilo-gradle-home-balance-jetbrains-ui ./gradlew --no-configuration-cache typecheck :frontend:test --tests ai.kilocode.client.settings.UserProfileConfigurableTest :backend:test --tests ai.kilocode.backend.cli.ApiModelSerializationTest :build-tasks:test --tests normalization.OpenApiSpecNormalizerTest`
- Agent: `env TMPDIR=/private/tmp JAVA_HOME=/opt/homebrew/opt/openjdk@21 GRADLE_USER_HOME=/private/tmp/kilo-gradle-home-balance-jetbrains-ui bun run typecheck` from `packages/kilo-jetbrains`
- Agent launched the JetBrains sandbox with `runIde`
- Tested the changes in the Jetbrain extension launched

### Reviewer test steps

1. Sign in with a personal Kilo account
2. Open the JetBrains profile settings and confirm balance, Kilo Pass usage, bonus credits, renewal date, Dashboard, and Top up actions render.
3. Switch to an org account and confirm Kilo Pass details are hidden.

### Blocked checks and substitute verification

- The root pre-push hook initially failed because the local default `~/.gradle` cache resolved corrupt plugin artifacts. The branch was pushed with `--no-verify` after the same JetBrains typecheck passed with a clean `GRADLE_USER_HOME`.

## Checklist

- [x] Issue linked above, or exception explained
- [x] Tests/verification described
- [x] Screenshots/video included for visual changes, or marked N/A
- [x] Changeset considered for user-facing changes
- [x] I personally reviewed the diff and can explain the changes, including any AI-assisted work.

### CODE REVIEW

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/BalanceFormat.kt [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
+import java.time.Instant
+import java.time.ZoneId
+import java.time.format.DateTimeFormatter
+import java.util.Locale
 
-private val FMT = DecimalFormat("\$#,##0.00")
+private val SYMBOLS = DecimalFormatSymbols(Locale.US)
+private val FMT = DecimalFormat("\$#,##0.00", SYMBOLS)
+private val SHORT = DecimalFormat("\$#,##0", SYMBOLS)
+private val DATE = DateTimeFormatter.ofPattern("MMM d", Locale.US).withZone(ZoneId.systemDefault())
```
[the remark this answers is not in the document]
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/BalanceFormat.kt]: Addressed

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 2]
```diff
@@ -178,11 +244,68 @@ internal class LoggedInProfileUi(
                 changed = true
             }
         }
+        changed = syncPass(if (personal) profile.kiloPass else null, bal != null && personal) || changed
```
[the remark this answers is not in the document]
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt]: Addressed

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 3]
```diff
@@ -161,26 +226,87 @@ internal class LoggedInProfileUi(
         if (showEmail && emailLabel.text != profile.email) emailLabel.text = profile.email
 
         val bal = profile.balance
+        val personal = profile.currentOrgId == null
+        val item = if (personal) profile.kiloPass else null
+        val showPass = personal && (bal != null || item != null)
         var changed = false
+        changed = visible(titleLabel, bal != null) || changed
+        changed = visible(valueLabel, bal != null) || changed
+        changed = visible(refreshBtn, bal != null) || changed
```
[the remark this answers is not in the document]
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt]: Addressed

**Review by kirillk — COMMENTED** [review]
Review comments left inline. Main asks are adding coverage for the new retained Swing meter/hidden-row states and tightening a couple JetBrains UI conformance/parity details.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 4]
```diff
@@ -161,26 +226,87 @@ internal class LoggedInProfileUi(
...
+        if (passInfo.isVisible != info) {
+            passInfo.isVisible = info
+            changed = true
+        }
+        if (item != null) {
+            val usage = KiloBundle.message(
+                "profile.pass.used",
+                formatShortBalance(item.currentPeriodUsageUsd),
+                formatShortBalance(item.currentPeriodBaseCreditsUsd),
+            )
```
Comment (kirillk) [inline packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:263]: This is the main new piece of UI state/paint logic, but the tests only assert that `passPanel` is visible. Please add coverage for the meter path itself: fraction clamping for usage > base, divide-by-zero/base-zero handling, and preferably the no-op repaint guard. The JetBrains AGENTS.md calls out retained Swing UI updates as needing tests against the actual implementation, and this is the highest-risk untested branch in the PR.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 5]
```diff
@@ -161,26 +226,87 @@ internal class LoggedInProfileUi(
...
-        } else {
-            if (balanceCard.isVisible) {
-                balanceCard.isVisible = false
+            val base = item.currentPeriodBaseCreditsUsd.coerceAtLeast(1.0)
+            passMeter.setFraction(item.currentPeriodUsageUsd / base)
+            val bonus = if (item.currentPeriodBonusCreditsUsd > 0.0) {
+                formatBalance(item.currentPeriodBonusCreditsUsd)
+            } else null
+            if (passBonusValue.text != bonus.orEmpty()) {
+                passBonusValue.text = bonus.orEmpty()
```
Comment (kirillk) [inline packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:274]: Please add UI coverage for the hidden-row cases here: `currentPeriodBonusCreditsUsd <= 0.0` should hide the Bonus row, and null/invalid `nextBillingAt` should hide the Renews row. Current tests only cover the positive bonus + valid renewal path, so regressions in these visibility branches would pass.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 6]
```diff
@@ -161,26 +226,87 @@ internal class LoggedInProfileUi(
...
+            val renewVisible = reset != null
+            if (passRenewRow.isVisible != renewVisible) {
+                passRenewRow.isVisible = renewVisible
+                changed = true
+            }
+        }
+        val link = show && item == null
+        if (passLink.isVisible != link) {
+            passLink.isVisible = link
+            changed = true
```
Comment (kirillk) [inline packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:296]: JetBrains AGENTS.md says every method that touches Swing components should be annotated with `@RequiresEdt`. This helper mutates `component.isVisible`; please annotate it. Same applies to `PassMeter.setFraction(...)` below since it calls `repaint()` and mutates paint state.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt [hunk 7]
```diff
@@ -161,26 +226,87 @@ internal class LoggedInProfileUi(
...
+        if (item != null) {
+            val usage = KiloBundle.message(
+                "profile.pass.used",
+                formatShortBalance(item.currentPeriodUsageUsd),
+                formatShortBalance(item.currentPeriodBaseCreditsUsd),
+            )
+            if (passUsageLabel.text != usage) {
+                passUsageLabel.text = usage
                 changed = true
             }
```
Comment (kirillk) [inline packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:267]: The VS Code/TUI balance display prefixes bonus credits with `+` (`"+" + format(...)`). This JetBrains UI shows the amount without the plus sign. If the intent is to match the VS Code balance UI, please either include the `+` here or call out why JetBrains should intentionally differ.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/BalanceFormat.kt [hunk 8]
```diff
@@ -1,8 +1,23 @@
...
+import java.util.Locale
 
-private val FMT = DecimalFormat("\$#,##0.00")
+private val SYMBOLS = DecimalFormatSymbols(Locale.US)
+private val FMT = DecimalFormat("\$#,##0.00", SYMBOLS)
+private val SHORT = DecimalFormat("\$#,##0", SYMBOLS)
+private val DATE = DateTimeFormatter.ofPattern("MMM d", Locale.US).withZone(ZoneOffset.UTC)
 
 /** Format a USD balance value for display (e.g. `$1,234.56`). */
 internal fun formatBalance(value: Double): String = FMT.format(value)
```
Comment (kirillk) [inline packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/BalanceFormat.kt:17]: Small parity issue: VS Code/TUI uses `Math.round` for the short Kilo Pass amounts, while `DecimalFormat("$#,##0")` defaults to HALF_EVEN rounding. They diverge on `.5` values (e.g. 2.5). If exact UI parity matters here, set `SHORT.roundingMode = RoundingMode.HALF_UP` and add a small test for that edge case.

Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/BalanceFormat.kt:17]: Great catch. I addressed in newest changes.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt (hunk 7 above)
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:267]: thanks for the callout. It was addressed now to match the VS Code UI

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt (hunk 6 above)
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:296]: Added. thank you

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt (hunk 4 above)
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:263]: thanks. Addressed in the last commit

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt (hunk 5 above)
Comment (johnnyeric) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/settings/profile/LoggedInProfileUi.kt:274]: thanks. Addressed in the last commit

**Review by kirillk — APPROVED** [review]
pls make sure tests pass
